### PR TITLE
implement equivalent sorts

### PR DESCRIPTION
**Pull request by osa1**
Very crude first implementation. Some tests are expected to fail. Not ready for merge.

@yilongli @andreistefanescu can you have a look?

### CODE REVIEW

File: k-core/src/main/java/org/kframework/utils/Poset.java [hunk 1]
```diff
@@ -2,19 +2,14 @@
...
-import java.util.Set;
-import java.util.Stack;
+import java.util.*;
 
 import org.apache.commons.collections4.set.UnmodifiableSet;
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ArrayTable;
 import com.google.common.collect.Table;
+import org.kframework.kil.Sort;
```
Comment (yilongli) [inline k-core/src/main/java/org/kframework/utils/Poset.java]: A `Poset` should be generic enough to have nothing to do with `Sort`.

Comment (osa1) [inline k-core/src/main/java/org/kframework/utils/Poset.java]: It's not used anywhere, just forgot to remove import statement.

File: k-core/src/main/java/org/kframework/utils/Poset.java [hunk 2]
```diff
@@ -22,6 +17,7 @@
 
     private java.util.Set<Tuple<T>> relations = new HashSet<>();
     private java.util.Set<T> elements = new HashSet<>();
+    private Map<T, Set<T>> equivalentElems = new HashMap<>();
```
Comment (yilongli) [inline k-core/src/main/java/org/kframework/utils/Poset.java:27]: an alternative is to have one representative for each equivalence class, that is, a map from T to T

File: k-core/src/main/java/org/kframework/utils/Poset.java [hunk 3]
```diff
@@ -2,19 +2,14 @@
...
 
 import java.io.Serializable;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
-import java.util.Stack;
+import java.util.*;
```
Comment (yilongli) [inline k-core/src/main/java/org/kframework/utils/Poset.java]: I suggest you set a larger limit in your IDE, we agreed in the past that we should not collapse import statements like these

**Comment by osa1** [comment]
This is good. As far as I can see there's only one failure and that's the problematic case I mentioned in #1011 . Current expected output of it is wrong.

File: k-core/src/main/java/org/kframework/compile/transformers/CompleteSortLattice.java [hunk 4]
```diff
@@ -92,20 +84,46 @@ public ASTNode visit(Module node, Void _)  {
...
-                    for (Production production : productions) {
+                    for (Production production : sort1Prods) {
                         if (!context.isSubsorted(sort2, production.getChildSort(0))) {
                             continue sort2_loop;
                         }
                     }
 
-                    transformedNode.addSubsort(sort2, sort1, context);
-                    change = true;
+                    if (context.isSubsorted(sort1, sort2)) {
```
Comment (yilongli) [inline k-core/src/main/java/org/kframework/compile/transformers/CompleteSortLattice.java:95]: This is not enough, I think. For example, you have `syntax A ::= X | Y` and `syntax KResult ::= X | Y`, you have to somehow compare the subsorts of A with subsorts of KResult to conclude that they are equivalent sorts. I don't think such information is already computed in context?

**Comment by yilongli** [comment]
The current regression test suite is not very strong. I suspect there are many places that use equivalent sorts. It would be nice to write a few tests about this.

File: k-core/src/main/java/org/kframework/compile/transformers/CompleteSortLattice.java (hunk 4 above)
Comment (osa1) [inline k-core/src/main/java/org/kframework/compile/transformers/CompleteSortLattice.java:95]: To be honest, I don't understand how subsorts are declared*. The removed code was declaring `sort1 < sort2` and I replaced that code with a code that assumes `sort1 < sort2`. So if it was previously correct, it should still be correct. (see also the else block, which is same as old code)

*: The code checks if `childNode`s of all productions of `sort1` are subsorts of `sort2` and if that's the case, declares `sort1 < sort2`. I don't know what's meant by `childNode` and it's not documented anywhere. I assume it's RHS of the production, and if that's really the case then looks correct to me.

**Comment by osa1** [comment]
I managed to generate a problematic definition file for this patch:

```
module TEST

  syntax S1 ::= Int
  syntax S2 ::= Bool | S1
  syntax KResult ::= S2
  syntax S3 ::= Int | Bool
  syntax KItem ::= "test"

endmodule
```

When I run `kompile --backend java` using Java 8, this is compiled in 20 seconds, but when I run same command using Java 7, it's getting stuck in an infinite loop(or taking huge amounts of time for some other reason -- killed the process after 114 seconds which is definitely not normal)

EDIT: I found the bug.

**Comment by osa1** [comment]
@yilongli @andreistefanescu I updated one test and created two new tests: https://github.com/osa1/k/commit/88d407ea71ab71b4add77f9369b8360959799ec0 https://github.com/osa1/k/commit/277819b199b8a079d43462dd5ff1125147fa9c41 can you have a look and tell me if outputs look correct?

File: k-core/src/main/java/org/kframework/utils/Poset.java [hunk 5]
```diff
@@ -47,7 +50,27 @@ public void addRelation(T big, T small) {
...
+        if (bigEqs == null) {
+            bigEqs = new HashSet<>();
+            bigEqs.add(big);
+        }
+        if (smallEqs == null) {
+            smallEqs = new HashSet<>();
+            smallEqs.add(small);
+        }
+
+        for (T s1 : bigEqs) {
```
Comment (yilongli) [inline k-core/src/main/java/org/kframework/utils/Poset.java:69]: This looks too expensive to me. Can you precompute some information and eliminate this nested loop? For example, have one representative for each equivalent class.

File: k-core/src/main/java/org/kframework/compile/transformers/CompleteSortLattice.java (hunk 4 above)
Comment (yilongli) [inline k-core/src/main/java/org/kframework/compile/transformers/CompleteSortLattice.java:95]: alright, then I am fine with it

**Comment by yilongli** [comment]
The tests look good to me.

File: k-core/src/main/java/org/kframework/utils/Poset.java (hunk 5 above)
Comment (andreistefanescu) [inline k-core/src/main/java/org/kframework/utils/Poset.java:69]: I agree with @yilongli . An alternative is to keep for each set of equivalent sorts a representative, and then have a `HashMap` from each sort to the representative of the equivalence class. The `relations` should only contain pairs of representatives of equivalence classes.

**Comment by osa1** [comment]
Found another bug with subsorting and added a test for that.

**Comment by osa1** [comment]
We need to handle transitive closure generation and cycle detection too.

**Pull request closed without merging**